sdk/storage/azure-storage-common/inc/azure/storage/common/paged_response.hpp

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <cstdlib>
+#include <string>
+
+#include <azure/core/context.hpp>
+#include <azure/core/http/raw_response.hpp>
+
+namespace Azure {
+
+template <class Derived> class PagedResponse {
+public:
+  std::string CurrentPageToken;
+  std::string NextPageToken;
+  std::unique_ptr<Azure::Core::Http::RawResponse> RawResponse;
+
+  bool HasMorePages() const { return m_hasMorePages; }
+
+  void MoveToNextPage(const Azure::Core::Context& context = Azure::Core::Context())
+  {
+    static_assert(
+        std::is_base_of<PagedResponse, Derived>::value,
+        "The template argument \"Derived\" should derive from PagedResponse<Derived>.");
+
+    if (NextPageToken.empty())
+    {
+      m_hasMorePages = false;
+      return;
+    }
+    CurrentPageToken = NextPageToken;
+    static_cast<Derived*>(this)->OnNextPage(context);
```

Comment:
`OnNextPage` will send HTTP requests, and may fail. Do you think it's valuable to keep current page unchanged if some error happens when trying to get next page?

Scenario:
```C++
auto page = client.ListSomething();

try {
    page.MoveToNextPage();
} catch (...) {}

// customers still have access to `page`, maybe they want get `NextPageToken` and do something about it.
```

@JeffreyRichter @ahsonkhan

Reply:
> Do you think it's valuable to keep current page unchanged if some error happens when trying to get next page?

Yes.

Would the fix just be to set `CurrentPageToken` after the call to `OnNextPage()`, or would we also have to create a back-up of `NextPageToken` first?
Either way, only updating this state after a successful completion of `OnNextPage()` is a good idea.

Reply:
I did it in a very simple way. Check it out https://github.com/Azure/azure-sdk-for-cpp/pull/2116/commits/c1af7fc1b2db09108204f0a7e23638c8c75e32cf.

Reply:
I don't understand how replacing `std::Move(NextPageToken)` with a direct assignment when setting `ContinuationToken` within `OnNextPage` avoids changing `CurrentPageToken`. On line 33 above, you already changed it to the next token. Now, if `OnNextPage` throws, the `CurrentPageToken` already changed to next. How does this achieve what you wanted?

Don't we want the following: If `MoveToNextPage()` throws, `CurrentPageToken` should remain what it  was prior to that method call.